# Importing Modules

In [14]:
import numpy as np
import pandas as pd

# Getting Required Data

In [16]:
!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

--2021-09-08 06:16:50--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1.4M) [text/plain]
Saving to: ‘data.csv’

data.csv            100%[===================>]   1.41M  --.-KB/s    in 0.09s   

2021-09-08 06:16:51 (16.3 MB/s) - ‘data.csv’ saved [1475504/1475504]



# Q1. Checking Numpy Version
[useful](https://stackoverflow.com/questions/1520234/how-do-i-check-which-version-of-numpy-im-using)




In [8]:
np.version

<module 'numpy.version' from '/usr/local/lib/python3.7/dist-packages/numpy/version.py'>

In [9]:
np.version.version

'1.19.5'

In [13]:
np.__version__

'1.19.5'

# Q2. Checking Pandas Version

In [15]:
pd.__version__

'1.1.5'

In [18]:
df = pd.read_csv("data.csv")

In [19]:
df.head()

Make       Model  Year  ... city mpg  Popularity   MSRP
0  BMW  1 Series M  2011  ...       19        3916  46135
1  BMW    1 Series  2011  ...       19        3916  40650
2  BMW    1 Series  2011  ...       20        3916  36350
3  BMW    1 Series  2011  ...       18        3916  29450
4  BMW    1 Series  2011  ...       18        3916  34500

[5 rows x 16 columns]

In [20]:
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [25]:
df['Make'] == "BMW"

0         True
1         True
2         True
3         True
4         True
         ...  
11909    False
11910    False
11911    False
11912    False
11913    False
Name: Make, Length: 11914, dtype: bool

In [27]:
bmw_data = df[df['Make'] == "BMW"]

In [28]:
bmw_data.describe()

Year   Engine HP  ...  Popularity           MSRP
count   334.000000  334.000000  ...       334.0     334.000000
mean   2014.347305  326.907186  ...      3916.0   61546.763473
std       4.387983  105.681659  ...         0.0   27982.682153
min    1995.000000  170.000000  ...      3916.0    4697.000000
25%    2015.000000  240.000000  ...      3916.0   42300.000000
50%    2016.000000  300.000000  ...      3916.0   51850.000000
75%    2016.000000  394.000000  ...      3916.0   79725.000000
max    2017.000000  600.000000  ...      3916.0  141200.000000

[8 rows x 8 columns]

In [31]:
df["Engine HP"].na

AttributeError: ignored

# Question 3

## What's the average price of BMW cars in the dataset?
* 61546.763473




In [34]:
bmw_data.MSRP.mean()

61546.76347305389

# Question 4 
## Number of missing values in "Engine HP" after 2015

*   69



In [37]:
df["Engine HP"].isnull().sum()

69

In [44]:
af_2015 = df[df["Year"] >= 2015]

In [45]:
af_2015["Engine HP"].isnull().sum()

51

# Question 5 
## Does the mean change after filling missing values?

*   



In [46]:
af_2015_eng_hp =  af_2015["Engine HP"]

In [47]:
af_2015_eng_hp.isnull().sum()

51

In [54]:
mean_before = round(af_2015_eng_hp.mean())
mean_before

273

In [56]:
af_2015_eng_hp = af_2015_eng_hp.fillna(mean_before)

In [57]:
mean_after = round(af_2015_eng_hp.mean())
mean_after

273

# Question 6

## Steps:
1.   Select all the "Rolls-Royce" cars from the dataset.
2.   List item
3.   Select only columns "Engine HP", "Engine Cylinders", "highway MPG".
4.   Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 7 rows).
5.   Get the underlying NumPy array. Let's call it X.
6.   Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
7. Invert XTX.
8. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 7 rows).

## Sum of elements of the inverse matrix

*  

In [63]:
rolls_royce = df[df["Make"] == "Rolls-Royce"]
len(rolls_royce)

31

In [65]:
rolls_royce_3c = rolls_royce[[ "Engine HP", "Engine Cylinders", "highway MPG"]]

In [71]:
X = rolls_royce_3c.drop_duplicates()
len(X)

7

In [80]:
x = X
x.shape

(7, 3)

In [79]:
x_t = X.T
x_t.shape

(3, 7)

In [82]:
x @ x_t

2921      3505      5275      5279      7443      7553      11448
2921   105914.0  183356.0  183386.0  183371.0  104971.0  147606.0  203211.0
3505   183356.0  317474.0  317512.0  317493.0  181715.0  255544.0  351855.0
5275   183386.0  317512.0  317554.0  317533.0  181745.0  255582.0  351897.0
5279   183371.0  317493.0  317533.0  317513.0  181730.0  255563.0  351876.0
7443   104971.0  181715.0  181745.0  181730.0  104053.0  146295.0  201387.0
7553   147606.0  255544.0  255582.0  255563.0  146295.0  205714.0  283215.0
11448  203211.0  351855.0  351897.0  351876.0  201387.0  283215.0  389961.0

In [75]:
np.matmul(X, X.T)

ValueError: ignored

# Questions 7 
## Normal equation

* 

